src/Bedrock.Framework.Experimental/Transports/WebSockets/WebSocketPayloadEncoder.cs

```diff
@@ -0,0 +1,152 @@
+        /// in-place to avoid allocations.
+        /// </remarks>
+        /// <param name="input">The sequence to mask or unmask.</param>
+        /// <param name="length">The length of the total payload remaining.</param>
+        /// <param name="useSimd">Whether or not to use SIMD accelerated masking.</param>
+        /// <param name="position">The position in the consumed portion of the sequence.</param>
+        /// <returns>The number of bytes consumed, which may be less than the number of total payload bytes.</returns>
+        public unsafe long MaskUnmaskPayload(in ReadOnlySequence<byte> input, ulong length, bool useSimd, out SequencePosition position)
+        {
+            var lengthRemaining = (long)length;
+            position = input.Start;
+
+            if(input.IsSingleSegment)
+            {
+                var bytesToRead = (int)Math.Min((long)lengthRemaining, input.First.Length);
+                MaskUnmaskSpan(input.FirstSpan, bytesToRead, useSimd);
+
+                position = new SequencePosition(input.First, bytesToRead);
+                return lengthRemaining - bytesToRead;
+            }
+
+            foreach (var memory in input)
+            {
+                var bytesToRead = (int)Math.Min((long)lengthRemaining, memory.Length);
+                MaskUnmaskSpan(memory.Span, bytesToRead, useSimd);
+
+                position = new SequencePosition(memory, bytesToRead);
+                lengthRemaining -= bytesToRead;
+            }
+
+            return lengthRemaining;
+        }
+
+        /// <summary>
+        /// Masks or unmasks a WebSocket payload span.
+        /// </summary>
+        /// <param name="span">The span to mask or unmask.</param>
+        /// <param name="bytesToRead">The number of bytes to read from the span.</param>
+        /// <param name="useSimd">Whether or not to use SIMD accelerated masking.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe void MaskUnmaskSpan(in ReadOnlySpan<byte> span, long bytesToRead, bool useSimd)
+        {
+            var maskingKey = _maskingKey;
+
+            fixed (byte* dataStartPtr = &MemoryMarshal.GetReference(span))
+            {
+                byte* dataPtr = dataStartPtr;
+                byte* dataEndPtr = dataStartPtr + bytesToRead;
+                byte* maskPtr = (byte*)&maskingKey;
+
+                //If there is at least one int of data available
+                if (dataEndPtr - dataStartPtr >= sizeof(int))
+                {
+                    //Start by aligning to an int boundary, makes aligned loads faster later
+                    while ((ulong)dataPtr % sizeof(int) != 0)
+                    {
+                        Debug.Assert(dataPtr < dataEndPtr);
+
+                        *dataPtr++ ^= maskPtr[_currentMaskIndex];
+                        _currentMaskIndex = (_currentMaskIndex + 1) & 3; //Modulus shortcut for power of 2 values for perf, same as _currentMaskIndex % 4
```

Comment:
When casted to an unsigned (uint, ulong) the JIT will employ the bit-hack for you, so the comment is redundant and the code expressive.

Reply:
I'm not against this change necessarily, but do we want to rely on the JIT for always doing the right thing when it can be easily expressed?

Reply:
> rely on the JIT for always doing the right thing

If the JIT won't do this optimization for us, then it would be a regression (which will be caught by automatic perf-testing).

> when it can be easily expressed

You and I am aware of this bit-trick. Many other will be. But is it as readable as `% 4`?
Where will we draw the line for "easily expressed"? 
If it is easily expressed why is the comment needed?

Reply:
When I said "easily expressed", I suppose I meant "can be expressed by the language", not so much in the sense of "easily understood by the general programming public". Nonetheless, your point definitely stands and I've changed it out.